src/combinatorics/simple_graph/inc_matrix.lean

```diff
@@ -0,0 +1,409 @@
+  { rintros ⟨e1, he1⟩ ⟨e2, he2⟩ he1' he2' hr,
+    ext,
+    simp only [subtype.mk_eq_mk] at hr,
+    simp only [hr] },
+  { rintros ⟨e, he⟩ he',
+    use [e],
+    { simpa only [true_and, finset.mem_univ, finset.mem_filter] using he'},
+    { simp only [finset.mem_univ, exists_prop_of_true] } }
+end
+
+namespace simple_graph
+
+universe v
+variables {V : Type v} [fintype V] (G : simple_graph V) (R) [decidable_rel G.adj] [decidable_eq V]
+
+-- ## Incidence matrix M
+
+/-- `inc_matrix G R` is the matrix `M` such that `M i e = 1` if vertex `i` is an
+endpoint of the edge `e` in the simple graph `G`, otherwise `M i j = 0`. -/
+def inc_matrix : matrix V G.edge_set R
+| i e := if (e : sym2 V) ∈ G.incidence_set i then 1 else 0
+
+@[simp]
+lemma inc_matrix_apply {i : V} {e : G.edge_set} :
+  G.inc_matrix R i e = if (e : sym2 V) ∈ G.incidence_set i then 1 else 0 := rfl
+
+lemma inc_matrix_def : G.inc_matrix R = λ i e, ite ((e : sym2 V) ∈ G.incidence_set i) 1 0 :=
+by { ext, simp only [inc_matrix_apply] }
+
+-- ### Relation between inc_matrix elements and incidence_set property
+
+@[simp]
+lemma inc_matrix_zero {i : V} {e : G.edge_set} : G.inc_matrix R i e = 0 ↔ e.val ∉ G.incidence_set i :=
+by simp only [inc_matrix, ite_eq_right_iff, subtype.val_eq_coe, ←decidable.not_imp_not,
+              forall_true_left, not_false_iff, one_ne_zero]
+
+@[simp]
+lemma inc_matrix_one {i : V} {e : G.edge_set} : G.inc_matrix R i e = 1 ↔ e.val ∈ G.incidence_set i :=
+by simp only [inc_matrix, ite_eq_left_iff, subtype.val_eq_coe, ←decidable.not_imp_not,
+              set.not_not_mem, forall_true_left, not_false_iff, zero_ne_one]
+
+-- ### One - zero properties
+
+@[simp]
+lemma inc_matrix_not_zero {i : V} {e : G.edge_set} : ¬ G.inc_matrix R i e = 0 ↔ G.inc_matrix R i e = 1 :=
+by simp only [inc_matrix_zero, inc_matrix_one, set.not_not_mem]
+
+@[simp]
+lemma inc_matrix_not_one {i : V} {e : G.edge_set} : ¬ G.inc_matrix R i e = 1 ↔ G.inc_matrix R i e = 0 :=
+by simp only [inc_matrix_zero, inc_matrix_one]
+
+lemma inc_matrix_zero_or_one {i : V} {e : G.edge_set} :
+  G.inc_matrix R i e = 0 ∨ G.inc_matrix R i e = 1 :=
+by { rw [inc_matrix_zero, inc_matrix_one], exact (em (e.val ∈ G.incidence_set i)).symm }
+
+@[simp]
+lemma inc_matrix_elements_product_one {i j : V} {e : G.edge_set} :
+  G.inc_matrix R i e * G.inc_matrix R j e = 1 ↔ G.inc_matrix R i e = 1 ∧ G.inc_matrix R j e = 1 :=
+begin
+  cases G.inc_matrix_zero_or_one R with H₀ H₁,
+  { rw H₀, simp only [if_t_t, mul_boole, inc_matrix_apply, zero_ne_one, false_and] },
+  { rw H₁, simp only [true_and, mul_boole, inc_matrix_apply, eq_self_iff_true] }
+end
+
+-- ### Helping lemmas for edges
+
+@[simp]
+lemma edge_val_equiv {e₁ e₂ : G.edge_set} : e₁.val = e₂.val ↔ e₁ = e₂ :=
+begin
+  split,
+  { exact subtype.eq },
+  { intro hyp,
+    rw hyp }
+end
+
+lemma edge_val_in_set {e : G.edge_set} : e.val ∈ G.edge_set :=
+by simp only [subtype.coe_prop, subtype.val_eq_coe]
```

Comment:
```suggestion
e.2
```

Reply:
Should I still keep the `edge_val_in_set` lemma or try to replace it with e.property where I can?